File: 323-number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        int components = n;
        UnionFind uf = new UnionFind(n);
        for (int[] edge : edges) {
            if (uf.union(edge[0], edge[1])) {
                components--;
            }
        }

        return components;
    }
}

class UnionFind {
    int[] parent;
    int[] size;

    public UnionFind(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int find(int x) {
        if (x == parent[x]) {
            return x;
        }
        return parent[x] = find(parent[x]);
    }

    public boolean union(int a, int b) {
        int root_a = find(a);
        int root_b = find(b);
        if (root_a == root_b) return false;
        if (size[root_a] > size[root_b]) {
            parent[root_b] = root_a;
            size[root_a] += size[root_b] + 1;
        } else {
            parent[root_a] = root_b;
            size[root_b] += size[root_a] + 1;
        }
        return true;
    }
}

File: 323-number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.py
class UnionFind:
    def __init__(self, n):
        self.rank = [0] * n
        self.parent = list(range(n))
    def find(self, a):
        if self.parent[a] != a:
            self.parent[a] = self.find(self.parent[a])
        return self.parent[a]
    def union(self, a, b):
        parent_of_a = self.find(a)
        parent_of_b = self.find(b)
        if parent_of_a == parent_of_b:
            return False
        if self.rank[parent_of_a] < self.rank[parent_of_b]:
            self.parent[parent_of_a] = parent_of_b
        elif self.rank[parent_of_a] > self.rank[parent_of_b]:
            self.parent[parent_of_b] = parent_of_a
        else:
            self.parent[parent_of_b] = parent_of_a
            self.rank[parent_of_a] += 1
        return True
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        uf = UnionFind(n)
        connections = 0
        for edge in edges:
            if uf.union(edge[0], edge[1]):
                connections += 1
        return n - connections